client/session.go

```diff
@@ -287,34 +321,200 @@ func (s *session) Write(id string, t time.Time, value float64, unit xtime.Unit,
+	if tsErr != nil {
+		return nil, tsErr
+	}
+
+	iters := s.seriesIteratorsPool.Get(len(ids))
+	iters.Reset(len(ids))
+
+	fetchBatchOpsByHostIdx = s.fetchBatchOpArrayArrayPool.Get()
+
+	s.RLock()
+	quorum = s.quorum
+	for idx := range ids {
+		idx := idx
+
+		var (
+			results  []m3db.Iterator
+			enqueued int
+			pending  int32
+			firstErr error
+			errs     int32
+		)
+
+		wg.Add(1)
+		completionFn := func(result interface{}, err error) {
+			resultsIdx := atomic.AddInt32(&pending, -1)
+			if err != nil {
+				n := atomic.AddInt32(&errs, 1)
+				if n == 1 {
+					firstErr = err
+				}
+			} else {
+				slicesIter := s.readerSliceOfSlicesIteratorPool.Get()
+				slicesIter.Reset(result.([]*rpc.Segments))
+				mixedIter := s.mixedReadersIteratorPool.Get()
+				mixedIter.Reset(slicesIter)
+				// Results is pre-allocated after creating fetch ops for this ID below
+				results[resultsIdx] = mixedIter
+			}
+			if resultsIdx != 0 {
+				// Requests still pending
+				return
+			}
+			if err := s.consistencyResult(quorum, enqueued, int(errs), firstErr); err != nil {
+				resultErrLock.Lock()
+				if resultErr == nil {
+					resultErr = err
+				}
+				resultErrs++
+				resultErrLock.Unlock()
+			} else {
+				iter := s.seriesIteratorPool.Get()
+				iter.Reset(ids[idx], startInclusive, endExclusive, results)
+				iters.SetAt(idx, iter)
+			}
+			wg.Done()
 		}
-		switch level {
-		case m3db.ConsistencyLevelAll:
-			return reportErr()
-		case m3db.ConsistencyLevelQuorum:
-			if success >= quorum {
-				// Meets quorum
-				break
+
+		if err := s.topoMap.RouteForEach(ids[idx], func(hostIdx int, host m3db.Host) {
+			// Inc safely as this for each is sequential
+			enqueued++
+			pending++
+
+			ops := fetchBatchOpsByHostIdx[hostIdx]
+
+			var f *fetchBatchOp
+			if len(ops) > 0 {
+				// Find the last and potentially current fetch op for this host
+				f = ops[len(ops)-1]
 			}
-			return reportErr()
-		case m3db.ConsistencyLevelOne:
-			if success > 0 {
-				// Meets one
-				break
+			if f == nil || f.Size() >= s.fetchBatchSize {
+				// If no current fetch op or existing one is at batch capacity add one
+				f = s.fetchBatchOpPool.Get()
+				fetchBatchOpsByHostIdx[hostIdx] = append(fetchBatchOpsByHostIdx[hostIdx], f)
+				f.request.RangeStart = rangeStart
+				f.request.RangeEnd = rangeEnd
+				f.request.RangeType = rpc.TimeType_UNIX_NANOSECONDS
+			}
+
+			// Append ID to this request
+			f.append(ids[idx], completionFn)
+		}); err != nil {
+			routeErr = err
+			break
+		}
+
+		// Once we've enqueued we know how many to expect so retrieve and set length
+		results = s.iteratorArrayPool.Get(enqueued)
+		results = results[:enqueued]
+	}
+
+	if routeErr != nil {
+		s.RUnlock()
+		return nil, routeErr
+	}
+
+	// Enqueue fetch ops
+	for idx := range fetchBatchOpsByHostIdx {
+		for _, f := range fetchBatchOpsByHostIdx[idx] {
+			if err := s.queues[idx].Enqueue(f); err != nil && enqueueErr == nil {
+				enqueueErr = err
```

Comment:
Do we want to break out of the loop early since we are going to return the error immediately down below anyway?